cmd/tegola/cmd/cache.go

```diff
@@ -279,35 +178,278 @@ var cacheCmd = &cobra.Command{
+	if err != nil || xi > maxXYatZ {
+		return 0, 0, 0, fmt.Errorf("invalid X value (%v)", xi)
+	}
+
+	yi, err := strconv.ParseUint(parts[f.Y], 10, 64)
+	if err != nil || yi > maxXYatZ {
+		return 0, 0, 0, fmt.Errorf("invalid Y value (%v)", yi)
+	}
+
+	return uint(zi), uint(xi), uint(yi), nil
+}
+
+func seedWorker(ctx context.Context, mt MapTile) error {
+	//	track how long the tile generation is taking
+	t := time.Now()
+
+	//	lookup the Map
+	m, err := atlas.GetMap(mt.MapName)
+	if err != nil {
+		return fmt.Errorf("error seeding tile (%+v): %v", mt.Tile, err)
+	}
+
+	z, x, y := mt.Tile.ZXY()
+
+	//	filter down the layers we need for this zoom
+	m = m.FilterLayersByZoom(x)
+
+	//	check if overwriting the cache is not ok
+	if !cacheOverwrite {
+		//	lookup our cache
+		c := atlas.GetCache()
+		if c == nil {
+			return fmt.Errorf("error fetching cache: %v", err)
+		}
+
+		//	cache key
+		key := cache.Key{
+			MapName: mt.MapName,
+			Z:       z,
+			X:       x,
+			Y:       y,
+		}
+
+		//	read the tile from the cache
+		_, hit, err := c.Get(&key)
+		if err != nil {
+			return fmt.Errorf("error reading from cache: %v", err)
+		}
+		//	if we have a cache hit, then skip processing this tile
+		if hit {
+			log.Infof("cache seed set to not overwrite existing tiles. skipping map (%v) tile (%v/%v/%v)", mt.MapName, z, x, y)
+			return nil
+		}
 	}
 
-	z, err := strconv.ParseUint(parts[0], 10, 64)
-	if err != nil || z > tegola.MaxZ {
-		return tile, fmt.Errorf("invalid Z value (%v)", z)
+	//	set tile buffer if it was configured by the user
+	if conf.TileBuffer > 0 {
+		mt.Tile.Buffer = float64(conf.TileBuffer)
 	}
 
-	maxXYatZ := maths.Exp2(z) - 1
+	//	seed the tile
+	if err = atlas.SeedMapTile(ctx, m, z, x, y); err != nil {
+		return fmt.Errorf("error seeding tile (%+v): %v", mt.Tile, err)
+	}
 
-	x, err := strconv.ParseUint(parts[1], 10, 64)
-	if err != nil || x > maxXYatZ {
-		return tile, fmt.Errorf("invalid X value (%v)", x)
+	//	TODO: this is a hack to get around large arrays not being garbage collected
+	//	https://github.com/golang/go/issues/14045 - should be addressed in Go 1.11
+	runtime.GC()
+
+	log.Infof("seeding map (%v) tile (%v/%v/%v) took: %v", mt.MapName, z, x, y, time.Now().Sub(t))
+
+	return nil
+}
+
+func purgeWorker(mt MapTile) error {
+
+	z, x, y := mt.Tile.ZXY()
+
+	log.Infof("purging map (%v) tile (%v/%v/%v)", mt.MapName, z, x, y)
+
+	//	lookup the Map
+	m, err := atlas.GetMap(mt.MapName)
+	if err != nil {
+		return fmt.Errorf("error seeding tile (%+v): %v", mt.Tile, err)
 	}
 
-	y, err := strconv.ParseUint(parts[2], 10, 64)
-	if err != nil || y > maxXYatZ {
-		return tile, fmt.Errorf("invalid Y value (%v)", y)
+	//	purge the tile
+	ttile := tegola.NewTile(mt.Tile.ZXY())
+	if err = atlas.PurgeMapTile(m, ttile); err != nil {
+		return fmt.Errorf("error purging tile (%+v): %v", mt.Tile, err)
+	}
+
+	return nil
+}
+
+func sendTiles(zooms []uint, c chan *slippy.Tile) error {
+	defer close(c)
```

Comment:
I know you were just copying what was already there, but I had some time to look at this function more carefully. 

This function is hard to read for two reasons:

1. The If-Then-Else block is too long and too deeply nested. 
2. At the end of the function, all you end up doing is returning `nil`, which is the reader has to remember what was going on and then scroll down to the end to find that out. This makes things very had to keep in one's head.

I would recommend changing this over to: 

https://play.golang.org/p/E7DPABYoVa4
or
https://play.golang.org/p/93R58CrBauS

I prefer the `switch` as I think it looks cleaner, and find it easier to read. The understanding with a switch statement is that each block is independent.